core/src/jsMain/kotlin/Peripheral.kt

```diff
@@ -128,56 +125,76 @@ public class JsPeripheral internal constructor(
     private val gatt: BluetoothRemoteGATTServer
         get() = bluetoothDevice.gatt!! // fixme: !!
 
-    private var connectJob: Deferred<Unit>? = null
-
-    private fun connectAsync() = scope.async(start = LAZY) {
-        logger.info { message = "Connecting" }
-        _state.value = State.Connecting.Bluetooth
-
+    private val connectJob: SharedRepeatableTask = scope.sharedRepeatableTask {
         try {
-            registerDisconnectedListener()
-
-            gatt.connect().await() // todo: Catch appropriate exception to emit State.Rejected.
-            _state.value = State.Connecting.Services
-
-            discoverServices()
-            onServicesDiscovered(ServicesDiscoveredPeripheral(this@JsPeripheral))
-            _state.value = State.Connecting.Observes
-            logger.verbose { message = "Configuring characteristic observations" }
-            observers.onConnected()
+            openConnection()
         } catch (t: Throwable) {
             logger.error(t) { message = "Failed to connect" }
-            disconnectGatt()
+            disconnectJob.launchAsync()
             throw t
         }
-
-        logger.info { message = "Connected" }
-        _state.value = State.Connected
     }
 
-    private fun closeConnection() {
-        observationListeners.clear()
-        disconnectGatt()
-        unregisterDisconnectedListener()
-        _state.value = State.Disconnected()
+    private val disconnectJob: SharedRepeatableTask = scope.sharedRepeatableTask {
+        connectJob.join()
+        closeConnection()
     }
 
     public override suspend fun connect() {
-        val job = connectJob ?: connectAsync().also { connectJob = it }
-        job.await()
+        connectJob.launchAsync().await()
     }
 
     public override suspend fun disconnect() {
-        job.cancelAndJoinChildren()
-        connectJob = null
-        disconnectGatt()
+        connectJob.cancel()
```

Comment:
With JS being single threaded, this probably isn't an issue; but generally I'd recommend we use `cancelAndJoin` here, to ensure that any connect cancellation completes before starting the disconnect process.

In other words, if this were multithreaded, then in our `openConnection` we may be performing some actions on non-suspend lines such as `registerDisconnectListener` and it could be called while the `disconnectJob` is running if we don't `join` the cancelling `connectJob`. Whereas `join`ing means that we'll wait til we hit a suspension point (to cooperatively cancel) before the `disconnectJob` is kicked off.

Reply:
Good catch, will make this change also.